### ISSUE TITLE
Fullscreen Patch 0.48.1 MacOS 10.13.4 

### ISSUE THREAD

**Issue by megrimm** [pure-data/pure-data#370]
Possibly to related to #353 

When putting a patch into fullscreen using the green fullscreen button on the upper left the fullscreen patch has a bar across the top the same size as the MacOS menubar and a bar across the bottom the same size and the dock. Also, when patching, the mouse pointer is 50+ px off from where it should be.  For example, rolling over an outlet only changes mouse state many px below object. This occurs anywhere in patch.

I looked into the "fullscreen-plugin" we had a few years ago and this creates the same behavior.

Looking here (https://wiki.tcl.tk/44444#pagetocdebf8a96) i altered the plugin to reflect suggested code and all seems to work fine. Code for fullscreen should be something to this effect although i am still getting error on fullscreen exit which i can look at in a few days:

```
package require Tcl 8.5
package require Tk
package require pdwindow 0.1
package require pd_menus 0.1

variable fullscreen_windows {}

proc go_fullscreen {window} {
	global fullscreen_windows
	set cfs [wm attributes $window -fullscreen]
	if { $::tcl_platform(os) eq "Darwin" } {
		if { $cfs == 0 } {
		    set savevar [wm geometry $window]
		           }
	wm withdraw $window
 	} 
	wm attributes $window -fullscreen [expr {1-$cfs}]
	if { $::tcl_platform(os) eq "Darwin" } {
      wm deiconify $window
      if { $cfs == 1 } {
        after idle [list wm geometry $window $savevar]
		}
	}
}

proc withdraw_fullscreen {window} {
	global fullscreen_windows
	set cfs [wm attributes $window -fullscreen]
	if { $::tcl_platform(os) eq "Darwin" } {
		if { $cfs == 0 } {
		    set savevar [wm geometry $window]
		           }
	wm withdraw $window
 	} 
}

proc toggle_fullscreen {window} {
	global fullscreen_windows
	set window [winfo toplevel $window]
	if { [winfo class $window] eq "PatchWindow"} {
		if { [lsearch $fullscreen_windows $window] == -1 } {
			go_fullscreen $window
		} else {
			withdraw_fullscreen $window
		}
	}
}

if { $::tcl_platform(os) eq "Darwin" } {
    bind all <Control-Shift-Key-F> {+toggle_fullscreen %W}
} else {
    bind all <F11> {+toggle_fullscreen %W}
}
```

**Comment by danomatika** [comment]
Pd doesn't know about the size of the menubar and the window handling was never really designed for the newer macOS "fullscreen" usage where the menubar & window chrome are hidden. With the move to TK 8.5, going fullscreen now works but as you see Pd doesn't know quite how to deal with it. I think the easiest solution is to disable this aspect fo fullscreen, otherwise we have to temporarily figure out which windows are fullscreen and then calculate the difference in offset (annoying).

**Comment by danomatika** [comment]
I've tried and there doesn't appear to be a way to check the fullscreen status of a window in Tk 8.5.19 on macOS. [wm $window attributes -fullscreen] always returns 0...

**Comment by megrimm** [comment]
yeah. just to log the issue this is what happens after exiting full screen using the modified code above:

```
can't read "savevar": no such variable
can't read "savevar": no such variable
    while executing
"list wm geometry $window $savevar"
    (procedure "go_fullscreen" line 14)
    invoked from within
"go_fullscreen $window"
    (procedure "toggle_fullscreen" line 6)
    invoked from within
"toggle_fullscreen .x100310d40.c"
    (command bound to event)
```

**Comment by danomatika** [comment]
>When putting a patch into fullscreen using the green fullscreen button on the upper left

The problem is that this button is provided by the OS and there is no way to override it unless changes are made *within* Tk. The plugin you're referencing tries to intercept the key command and adjust for this, but this will never be called if you just click the button and we cannot remove the button either. The best solution would be if Tk either sent an event when a window goes into fullscreen or at least set the window's -fullscreen attribute. It currently only does this is you set the attribute within Tk.

**Comment by danomatika** [comment]
Also, you can double click the patch titlebar and/or hold option while clicking the green button to Zoom the window, ie. grow to fill the screen without entering fullscreen mode.

**Comment by danomatika** [comment]
I see the Tk team are working on this upstream, so you might just need to wait on this. I tried hacking on the Tk source code and wasn't able to make any real improvements. Just double-click the titlebar...

**Comment by megrimm** [comment]
yeah the double clicking title bar thing i figured out the other day.

where is the tk discussion on this? i didnt find it.

I mean the fullscreen-plugin.tcl script works fine with modification other than the error on exit. maybe green button can be disabled until tk fix?

>> The problem is that this button is provided by the OS and there is no way to override it 

... but then im assuming this is not possible either.

**Comment by danomatika** [comment]
This is a duplicate of #353. Please move further discussion there. Reopen if I'm wrong.

**Comment by danomatika** [comment]
This should be fixed in 0.51-3 and the upcoming 0.51.4 as both now use Tk 8.6.10 which improves Tks usage of the Cocoa API.
